File: src/ax.c
#include <stdio.h>
  void cmxm(double *a, int n1, double *b, int n2, double *c, int n3) {
    double (*am)[n2][n1] = (double (*)[n2][n1]) a;
    double (*bm)[n3][n2] = (double (*)[n3][n2]) b;
    double (*cm)[n3][n1] = (double (*)[n3][n1]) c;

    int i,j,k;
    for (j = 0; j < n3; j++) {
      for (i = 0; i < n1; i++) {
        (*cm)[j][i] = 0.0;

        for (k = 0; k < n2; k++) {
          (*cm)[j][i] += (*am)[k][i]*(*bm)[j][k];
        }
      }
    }

    return;
  }


  void cadd2(double *a, double *b, int n) {

    int i;
    for (i=0; i < n; i++) {
      a[i] += b[i];
    }

    return;
  }


  void axcuda_(int *nx1,int *ny1,int *nz1,int *nelt,int *ldim,
              double w[*nelt][(*nx1)*(*ny1)*(*nz1)],
              double u[*nelt][(*nx1)*(*ny1)*(*nz1)],
              double g[*nelt][(*nx1)*(*ny1)*(*nz1)][2*(*ldim)],
              double dxm1[*nx1][*nx1], double dxtm1[*nx1][*nx1]) {

    int nxyz = (*nx1)*(*ny1)*(*nz1);
    int m1 = (*nx1);
    int m2 = m1*m1;
    int m3 = m1*m1*m1;

    double ur[nxyz],us[nxyz],ut[nxyz];

    int e, k, i;
    for (e = 0; e < *nelt; e++) {
      cmxm((double *) dxm1, m1, &(u[e][0]), m1, (double *) ur, m2);
      for (k = 0; k < *nz1; k++) {
        cmxm((double*) &(u[e][(*nx1)*(*ny1)*k]), m1, (double*) dxtm1, m1, &us[(*nx1)*(*ny1)*k], m1);
      }
      cmxm(&(u[e][0]), m2, (double*)dxtm1, m1, (double*)ut, m1);

      for (i = 0; i < nxyz; i++) {
        double wr = g[e][i][0]*ur[i] + g[e][i][1]*us[i] + g[e][i][2]*ut[i];
        double ws = g[e][i][1]*ur[i] + g[e][i][3]*us[i] + g[e][i][4]*ut[i];
        double wt = g[e][i][2]*ur[i] + g[e][i][4]*us[i] + g[e][i][5]*ut[i];
        ur[i] = wr;
        us[i] = ws;
        ut[i] = wt;
      }

      cmxm((double*)dxtm1, m1, (double*)ur, m1, &(w[e][0]), m2);

      double wa[*nz1][*ny1][*nx1];
      for (k = 0; k < *nz1; k++) {
        cmxm(&us[(*nx1)*(*ny1)*k], m1, (double*)dxm1, m1, &(wa[k][0][0]), m1);
      }

      cadd2(&(w[e][0]), (double*)wa, m3);
      cmxm((double*)ut, m2, (double*)dxm1, m1, (double*)wa, m1);
      cadd2(&(w[e][0]), (double*)wa, m3);

    }

    return;
  }

File: src/ax.f
c-------------------------------------------------------------------------
c     subroutine axcuda(w,u,g,dxm1,dxtm1,nelt,ldim,nx1,ny1,nz1) ! Local matrix-vector product
      subroutine axcuda(nx1,ny1,nz1,nelt,ldim,w,u,g,dxm1,dxtm1) ! Local matrix-vector product

      implicit none

      integer nelt, ldim, nx1, ny1, nz1
      real w(nx1*ny1*nz1,nelt),u(nx1*ny1*nz1,nelt)
      real g(2*ldim,nx1*ny1*nz1,nelt)
      real dxm1(nx1,nx1), dxtm1(nx1,nx1)

      integer e, i, n, nxyz, m1, m2, m3, k
      real ur(nx1*ny1*nz1),us(nx1*ny1*nz1),ut(nx1*ny1*nz1)
      real wr, ws, wt
      real wa(nx1,ny1,nz1)
     
      nxyz = nx1*ny1*nz1
      n    = nx1-1
      m1 = n+1
      m2 = m1*m1
      m3 = m1*m1*m1

      do e=1, nelt
        call mxm(dxm1,m1,u(1,e),m1,ur,m2)
        do k=1,nz1
          call mxm(u(nx1*ny1*(k-1)+1,e),m1,dxtm1,m1,
     c             us(nx1*ny1*(k-1)+1),m1)
        enddo
        call mxm(u(1,e),m2,dxtm1,m1,ut,m1)

        do i=1,nxyz
           wr = g(1,i,e)*ur(i) + g(2,i,e)*us(i) + g(3,i,e)*ut(i)
           ws = g(2,i,e)*ur(i) + g(4,i,e)*us(i) + g(5,i,e)*ut(i)
           wt = g(3,i,e)*ur(i) + g(5,i,e)*us(i) + g(6,i,e)*ut(i)
           ur(i) = wr
           us(i) = ws
           ut(i) = wt
        enddo

        call mxm(dxtm1,m1,ur,m1,w(1,e),m2)

        do k=1,nz1
          call mxm(us(nx1*ny1*(k-1)+1),m1,dxm1,m1,wa(1,1,k),m1)
        enddo
        call add2(w(1,e),wa,m3)

        call mxm(ut,m2,dxm1,m1,wa,m1)
        call add2(w(1,e),wa,m3)
      end do

      return
      end
